history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
+	if nextEp < h.NextEpoch() {
+		return ErrBadEpoch
+	}
+	nextStr := h.currentSTR.generateNextSTR(m, nextEp)
+	h.currentSTR = nextStr
+	return nil
+}
+
+func (h *History) LookUp(key string) (MerkleNode, *AuthenticationPath) {
+	str := h.currentSTR
+	return str.tree.Get(key)
+}
+
+func (h *History) LookUpInEpoch(key string, ep int64) (MerkleNode, *AuthenticationPath) {
+	str := h.GetSTR(ep)
+	return str.tree.Get(key)
+}
+
+func (h *History) GetSTR(ep int64) *SignedTreeRoot {
+	pointer := h.currentSTR
+	for pointer.epoch > ep && pointer != nil {
+		if pointer.prev == nil {
+			return nil
+		}
+		pointer = pointer.prev
+	}
+	return pointer
+}
+
+func (h *History) NextEpoch() int64 {
+	if h.currentSTR == nil {
+		return h.epochInterval
+	}
+	return h.epochInterval + h.currentSTR.epoch
+}
+
+func (h *History) MerkleTree() *MerkleTree {
+	return h.tree
+}
+
+func (h *History) NewTemporaryBinding(index, value []byte) []byte {
+	str := h.currentSTR
+	tb := getSTRBytesForSig(str.tree, str.epoch, str.prevEpoch, str.prevStrHash)
+	tb = append(tb, index...)
+	tb = append(tb, value...)
+	return crypto.Sign(str.key, tb)
+}
```

Comment:
I think you should remove this from this PR.  We can review it with the rest of #13 when that happens.  Let's try to keep the PRs as small as possible to ease the cognitive burden.